engine/src/main/java/org/hibernate/search/reader/impl/ManagedMultiReader.java

```diff
@@ -25,15 +32,94 @@
+	private ManagedMultiReader(IndexReader[] subReaders, IndexReader[] readersForClosing, ReaderProvider[] managers) throws IOException {
 		// If this flag isn't set to true, the MultiReader will increase the usage counter!
 		super( subReaders, true );
-		this.subReaders = subReaders;
+
+		this.readersForClosing = readersForClosing;
 		this.managers = managers;
-		assert subReaders.length == managers.length;
+		assert readersForClosing.length == managers.length;
+	}
+
+	public static ManagedMultiReader createInstance(IndexReader[] subReaders, ReaderProvider[] managers, Sort sort) throws IOException {
+		IndexReader[] effectiveReaders = getEffectiveReaders( subReaders, sort );
+		return new ManagedMultiReader( effectiveReaders, subReaders, managers );
+	}
+
+	/**
+	 * Gets the readers to be effectively used. Will be the given readers, if there is no sort involved. Otherwise each
+	 * directory reader will be wrapped in a {@link UninvertingReader} configured in a way to satisfy the requested
+	 * sorts.
+	 */
+	private static IndexReader[] getEffectiveReaders(IndexReader[] subReaders, Sort sort) {
+		if ( sort == null || sort.getSort().length == 0 ) {
+			return subReaders;
+		}
+
+		IndexReader[] uninvertingReaders = new IndexReader[subReaders.length];
+		Map<String, Type> mappings = getMappings( sort );
+
+		int i = 0;
+		for ( IndexReader reader : subReaders ) {
+			if ( reader instanceof DirectoryReader ) {
+				DirectoryReader directoryReader = (DirectoryReader) reader;
+
+				try {
+					uninvertingReaders[i] = UninvertingReader.wrap( directoryReader, mappings );
+				}
+				catch (IOException e) {
+					throw log.couldNotCreateUninvertingReader( directoryReader, e );
+				}
+			}
+			else {
+				log.readerTypeUnsupportedForInverting( reader.getClass() );
+				uninvertingReaders[i] = reader;
+			}
+
+			i++;
+		}
+
+		return uninvertingReaders;
+	}
+
+	/**
+	 * Returns the uninverting reader mappings required for the given non-null sort.
+	 */
+	private static Map<String, UninvertingReader.Type> getMappings(Sort sort) {
+		Map<String,UninvertingReader.Type> mappings = new HashMap<>();
+
+		for ( SortField sortField : sort.getSort() ) {
+			if ( sortField.getField() != null ) {
+				switch ( sortField.getType() ) {
+					case INT: mappings.put( sortField.getField(), Type.INTEGER );
+					break;
+					case LONG: mappings.put( sortField.getField(), Type.LONG );
+					break;
+					case FLOAT: mappings.put( sortField.getField(), Type.FLOAT );
+					break;
+					case DOUBLE: mappings.put( sortField.getField(), Type.DOUBLE );
+					break;
+					case STRING:
+					case STRING_VAL: mappings.put( sortField.getField(), Type.SORTED );
+					break;
+					case BYTES: mappings.put( sortField.getField(), Type.BINARY );
+					break;
+					case CUSTOM: // Nothing to do; expecting doc value fields created by the user
+					break;
+					default: log.sortFieldTypeUnsupported( sortField.getField(), sortField.getType() );
+				}
```

Comment:
Also, bridges are being ignored. My entity can have an ```int``` field and also use ```org.hibernate.search.bridge.builtin.IntegerBridge``` and it will end up being not stored as a ```NumericField```: in this case the Uninverter fails